Allow registration of custom MapGenStructures

maxanier opened the pull request:
Replaces #4496.
Instead of posting events at the appropriate positions this PR adds a way to add additional MapGenStructures to the ChunkGenererators which are then treated in the same way the vanilla ones (Village, ...).

This approach is more complex as the event based one, but it is cleaner and does not post events frequently.

## General Concept
Each standard (vanilla) ChunkGenerator (Overworld, Hell, End) fires an InitStructuresGenEvent during construction. Mods can register their additional structures to the event. Each ChunkGenerator has it's one MapGenStructureManager which holds the registered structures and manages the method calls.

## Spawning creatures
Since some structures would like to spawn creatures within their area, but there is no method in MapGenStructure for that (only in some subclasses). I added an interface which has to be implemented by any structures planning to spawn creatures.
The interface has two seperate methods shouldSpawn and getSpawns which is slightly redundant, but as there are three possible options here (custom list, empty list or biome list) I think this is the cleanest solution. Alternatively we could do
```
List list = structureManager.getSpawns();
return list ==null? biome.getSpawns : list;
```
But I don't feel comfortable with allowing having both null and an empty list as return value.

## Locate
Registered structures are automatically added to the locate command autocomplete, but the autocomplete is not dimension/generator specific (vanilla isn't either) since this would require more patches

## Test mod
It is probably not a good idea to include a "real" structure as test mod, so there is a dummy structure (for each dimension) which makes sure all methods are called. It is always "generated" in chunk x=4,z=4. See test mod for details

maxanier committed:
Allow registration of custom MapGenStructures

maxanier commented:
As usual if you can think of better names for the classes/fields let me know.

maxanier committed:
Added two comments

HenryLoenwind commented:
what about

```
List list = structureManager.getSpawns(biome.getSpawns);
return list;
```

?

liach commented on patches/minecraft/net/minecraft/world/gen/ChunkGeneratorEnd.java.patch:
```diff
@@ -25,26 +25,35 @@
 +        this.field_73212_b = ctx.getScale();
 +        this.field_185973_o = ctx.getIsland();
 +        this.field_185972_n = (MapGenEndCity) net.minecraftforge.event.terraingen.TerrainGen.getModdedMapGen(this.field_185972_n, net.minecraftforge.event.terraingen.InitMapGenEvent.EventType.END_CITY);
++        this.structureManager = net.minecraftforge.event.terraingen.TerrainGen.getAddedStructures(ChunkGeneratorHell.class);
```
1. The arguments seems incorrect.
2. Can you simply leave it in the end of the constructor of the root class?

maxanier replied:
1) Yes
2) Using this.getClass() ?

LexManos replied:
Why not just this?

maxanier replied:
Wasn't sure if it is a good idea to pass a "unfinished" object here. Modders could potentially call methods on this object and cause trouble. Not sure.
Also I couldn't use the generic event anymore.

But if you want I can change this

liach commented on patches/minecraft/net/minecraft/world/gen/ChunkGeneratorEnd.java.patch:
```diff
@@ -55,19 +64,56 @@
++
++    /* ======================================== FORGE START =====================================*/
++
++    private final net.minecraftforge.common.MapGenStructureManager structureManager;
```
Can you leave it protected for inheritance?

maxanier replied:
Sure

liach commented on src/main/java/net/minecraftforge/common/MapGenStructureManager.java:
```diff
@@ -0,0 +1,173 @@
+    /**
+     * @return A new array containing both the given strings as well as the registered structure names
+     */
+    public static String[] appendStructureNames(String[] vanilla)
```
Why?

maxanier replied:
For the locate command. I didn't want to move the vanilla hardcoded values here (if vanilla adds more structures to their list, it's probably easy to miss that when updating Forge if these valuse are hidden here.

Not a very nice implementation, I agree. Let me know if you have an alternative suggestion

liach commented on src/main/java/net/minecraftforge/common/MapGenStructureManager.java:
```diff
@@ -0,0 +1,173 @@
+    {
+        for (MapGenStructure struct : structures)
+        {
+            if (struct.getStructureName().equals(structureName))
```
I suggest a name to structure map to speed up lookup

liach replied:
With the map, you can handle the lookup of strings faster, too.

maxanier replied:
I tend to follow the vanilla  structure when doing patches/additions and don't think about alternatives, but a hashmap is a better solution.

liach commented on src/main/java/net/minecraftforge/event/terraingen/InitStructureGensEvent.java:
```diff
@@ -0,0 +1,58 @@
+
+    public List<MapGenStructure> getStructuresImmutable()
+    {
+        return ImmutableList.copyOf(structureList);
```
You can just returned a `Collections.unmodifiableList` to avoid extra copying.

maxanier replied:
Yes
I keep loosing track of all these "helper" methods :see_no_evil:

liach commented on src/test/java/net/minecraftforge/debug/StructureGenTest.java:
```diff
@@ -0,0 +1,216 @@
+            return super.generateStructure(worldIn, randomIn, chunkCoord);
+        }
+
+        @Nullable @Override
```
Annotations on two lines please

maxanier replied:
ok

liach commented on src/test/java/net/minecraftforge/debug/StructureGenTest.java:
```diff
@@ -0,0 +1,216 @@
+        }
+
+        @Override
+        public synchronized boolean generateStructure(World worldIn, Random randomIn, ChunkPos chunkCoord)
```
why `synchronized`?

maxanier replied:
It is synchronized in MapGenStructure

maxanier commented:
@HenryLoenwind good point, I will try to implement that one

maxanier commented:
@liach Thank you for reviewing. I will implement your suggestions later today

maxanier committed:
Implemented suggestions

maxanier commented:
Implemented your suggestions.

liach commented on src/test/java/net/minecraftforge/debug/StructureGenTest.java:
```diff
@@ -173,18 +182,18 @@ public boolean isInsideStructure(BlockPos pos)
 
         @Override
-        public List<Biome.SpawnListEntry> getSpawns(World world, EnumCreatureType creatureType, BlockPos pos)
+        public List<Biome.SpawnListEntry> getSpawns(List<Biome.SpawnListEntry> defaults, World world, EnumCreatureType creatureType, BlockPos pos)
```
You don't need to return a list any more; modification can be done within the list provided.

HenryLoenwind replied:
really?

```
defaults.clear();
defaults.addAll(myNeverChangingFinalStaticList);
```

all the time?

liach replied:
What if the user returns an immutable list?

maxanier replied:
ok

maxanier replied:
Yes, after implementing these changes and adding entity spawns to it, I agree with @HenryLoenwind .
It might be a nicer solution, if you don't change the spawn list every time, but since you most certainly always want to return your own list if you are implementing this interface.

Not sure about the following scenario though:
You want to add additional entries to the biome's ones.

With liach's approach you could just add them to the list. If we return a list you would have to create a new list to join the passed default list (which must not be modified) and your own final list.
Wouldn't make a difference in the background though as we  always have to create a new list in liach's approach anyway.

maxanier replied:
Any other opinions on this or should I just revert the change?
@liach

maxanier replied:
reverted for now

liach replied:
I originally thought both are O(N) and have the same amortized time, it is okay to be like this. Do we allow modifying the default list? If yes, ask users to return a copied new list every time.

maxanier replied:
The default list must not be modified because it is the list stored in the biome object
Therefore I am passing an unmodifiable list as default list here https://github.com/MinecraftForge/MinecraftForge/pull/4608/files#diff-a34fd8258a2226b73fa9f9405d45f27eR123

This is indirectly a change to the IChunkGenerator because in vanilla it always returns a modifiable list (as far as I can tell), but it definitively should not be modified so I think this is less of an issue. We could always ask for a copy of course, but I don't think that is necessary.

liach commented on src/main/java/net/minecraftforge/common/MapGenStructureManager.java:
```diff
@@ -157,17 +148,12 @@ public void generate(World worldIn, int x, int z, ChunkPrimer primer)
          */
         @Nonnull
-        List<Biome.SpawnListEntry> getSpawns(World world, EnumCreatureType creatureType, BlockPos pos);
+        List<Biome.SpawnListEntry> getSpawns(List<Biome.SpawnListEntry> defaults, World world, EnumCreatureType creatureType, BlockPos pos);
```
````java
void modifySpawns(List<Biome.SpawnListEntry> defaults, World world, EnumCreatureType creatureType, BlockPos pos);
````

williewillus commented on src/main/java/net/minecraftforge/common/MapGenStructureManager.java:
```diff
@@ -0,0 +1,159 @@
+ */
+public class MapGenStructureManager
+{
+    private static final List<String> allStructureNames = Lists.newArrayList();
```
this can be a set

maxanier replied:
indeed

williewillus commented on src/main/java/net/minecraftforge/common/MapGenStructureManager.java:
```diff
@@ -0,0 +1,159 @@
+     */
+    public static String[] appendStructureNames(String[] vanilla)
+    {
+        return ArrayUtils.addAll(vanilla, allStructureNames.toArray(new String[0]));
```
should we sort it so any command tab completes, etc. don't just return a registration-order jumbled mess?

maxanier replied:
maybe a good idea, but then I would stick with a list and sort the names upon registration

maxanier replied:
Sorting the allStructuresNames list in the MapGenStructureManager constructor now. toArray should keep the order.
This way the vanilla order is not influenced, but the additional ones are not based on registration order

maxanier replied:
Is this what you had in mind?

williewillus commented on src/test/java/net/minecraftforge/debug/StructureGenTest.java:
```diff
@@ -0,0 +1,205 @@
+            logger.info("Getting nearest structure {}", generator);
+            if (isComplete())
+            {
+                FMLLog.bigWarning("Structure Gen Test in {} successful", generator);
```
use info if it's not an error

maxanier replied:
but it is inside a (disabled by default) testmod and usually only triggered by a command, so I think that isn't that big a deal and it is much easier to spot during testing

mezz replied:
It's just bad practice, people use the test mods as coding examples sometimes.
I'll definitely see an info log if I'm looking for it, no problem.

maxanier replied:
Agreed, changed

maxanier committed:
Implemented more suggestions

maxanier committed:
Fixed spawn test

williewillus commented on src/main/java/net/minecraftforge/common/MapGenStructureManager.java:
```diff
@@ -152,6 +152,8 @@ public void generate(World worldIn, int x, int z, ChunkPrimer primer)
         /**
          * Collect the list of creatures of the given type that should spawn within the structures bounding box.
          * <p>
+         * If this method is called repeatedly you should always return exactly the same (equals) objects otherwise the spawning will not work.
```
why? is the structure not allowed to vary spawns based on e.g. time of day?

maxanier replied:
@williewillus The worldentityspawner calls this method twice.
1) to get possible spawns
2) to check if the selected spawn entry is allowed to spawn here
The entry selected after the first call has to be contained (list#contains) in the result of the second call otherwise the entity won't spawn.

But you can vary this over time of course.
I will try to clarify this comment

maxanier committed:
Address issues

maxanier commented:
- Reverted list related change.
I think the conversation got lost in an outdated diff https://github.com/MinecraftForge/MinecraftForge/pull/4608#discussion_r158140058

- Changed log
- Tried to clarify comment

maxanier commented:
@mezz How does this look, anything to improve?
@LexManos is this what you had in mind?  https://github.com/MinecraftForge/MinecraftForge/pull/4496#issuecomment-352226340

maxanier commented:
Also verified this with an actual example in a mod of mine.

pau101 commented:
@maxanier I would like to see support for the retro-generation of structures, hooking into `generateStructures(Chunk,int,int)` to be able to easily have the associated ocean monument behavior.

Also, I would find it helpful if `InitStructureGensEvent` were to include the IChunkGenerator instance it is for to simplify usage of the generator to fill a chunk primer, as seen in MapGenEndCity and WoodlandMansion.

Currently you have `allStructureNames` being manually sorted during every instantiation of MapGenStructureManager, to be a bit cleaner I'd advise using a TreeSet, which would also address williewillus's comment about using a set. Following that, simplify `addStructureName` by removing the contains check.

mezz commented:
This looks fairly good to me but I would like to see some more feedback from the community since it is a fairly large feature.

@Ivorforce can you take a look at this PR? I assume it would be useful for your mods.

Gegy commented on patches/minecraft/net/minecraft/world/gen/ChunkGeneratorEnd.java.patch:
```diff
@@ -25,26 +25,35 @@
+         if (this.field_73229_q)
+         {
+             this.field_185972_n.func_186125_a(this.field_73230_p, p_185932_1_, p_185932_2_, chunkprimer);
++            this.structureManager.generate(this.field_73230_p, p_185932_1_, p_185932_2_, chunkprimer);
```
Having `generate` and `generateStructure` could be a bit confusing -- maybe go with `primeStructure` and `generateStructure` instead?

maxanier replied:
Just sticking to the Vanilla/MCP names which are a little bit confusing here. The structure manager  `generate` calls the added structures `generate`, same for the `generateStructure`.
So I think it makes sense this way.

I agree that generate could be renamed to avoid confusion, but I would suggest renaming the MapGenStructure method first

maxanier committed:
Use TreeSet instead of manually sorted list.
Provided generator instance in event

maxanier commented:
@pau101 Good suggestions. Thank you. I switched to the TreeSet and added the generator instance to the object.
The generator object is not fully instantiated at that moment (as the event is posted from within the constructor). It is, however, at the very end, so I don't think this should cause any issues. Does it? I could add a warning to the event getter anyway?

About the retrogeneration. Yes. Just have to think about the best way to do this, as we probably don't want to call this for every MapGenStructure, but only per opt-in.

My initial idea is to create an empty interface which serves as an instance-of opt-in. The interface could also have a dedicated method to separate between the different calls (and include the chunk inhabitant time as argument).
What do you think?

pau101 commented:
The chunk generator not being fully constructed should be what is expected for a structure generator at this time, as the usage of it is intended to be storage for later reference when itself is performing generation. I would however still see a benefit in the addition of a comment on the getter noting that.

The interface design for retrogen is fine by me, but I would like the chunk provided and not only the inhabited time, as I could imagine other information being useful, especially chunk capabilities.

pau101 commented on src/main/java/net/minecraftforge/common/MapGenStructureManager.java:
```diff
@@ -0,0 +1,159 @@
+    {
+        for (MapGenStructure structure : structureMap.values())
+        {
+            if (structure instanceof ISpawningStructure && structure.isInsideStructure(pos))
```
The usage of `isInsideStructure` should be `isPositionInStructure` here, the former is used to check if a position is in one of the components of a structure, currently only used by the advancement location predicate. While the latter is for determining if a position is eligible for spawning, with the default behavior being check if the position is within the surrounding bounds of a structure.

maxanier replied:
Yes, thank you. Changed. Vanilla seems to use a different kind of check for every Vanilla spawning structure though, slightly strange.

Had to manipulated the StructureTest MapGenStructure slightly.  isPositionInStructure is always false here because we are generating an empty/dummy structure.

jabelar commented:
I'm a little confused as to the purpose of this. I successfully already register my structures by using the MapGenStructureIO.registerStructure() and MapGenStructureIO.registerStructureComponent() methods. For example I can add a new house type to vanilla villages or a whole new village. Then I can easily just replace the villageGenerator in the chunk generator.

I feel I have full control over the map gen structures with just a couple lines of code. What problem is this trying to solve?

EDIT: Okay I see that you want to add additional generators rather than replace/modify the existing ones. I normally work around this by replacing one of the vanilla generators with a "meta generator" -- i.e. there is nothing stopping the village generator from generating more than villages. But my approach isn't really mod-interworking friendly so your idea would be useful.

I'll check your implementation and review when I get a bit of time (maybe today).

jabelar commented:
Okay, just thought about this some more. Isn't the IWorldGenerator interface already intended for this? I realize the naming makes it confusing, but it is basically an interface already implemented by Forge to register map gen type things. The interface takes in parameters such as the IChunkGenerator so has all the information needed to augment the existing chunk generators.

Does IWorldGenerator interface already cover this?

pau101 commented:
@jabelar for full structure support there are a handful of methods which a custom structure needs ties into, which is what this pr is doing. For example the locate command, cartographer maps, natural mob spawning, and usage in the location predicate for advancement triggers. Doing a wrap of an existing generator like you mention only enables generation of your own structure but it can not support the aforementioned additional features.

An `IWorldGenerator` is not correct for structure generation, structures with save dava should be happening in two stages, first stage being identification of structure locations when a chunk generates, and second at the start of chunk population before biome decoration to place the blocks of the structure which exist in that chunk. An `IWorldGenerator` would be getting called only at the end of chunk population and you would also need to handle if structures should be generating. In the case for `ChunkGeneratorOverworld` in regards to `mapFeaturesEnabled` which require AT or reflection, but more importantly the chunk generator itself, particularly a modded one, should have complete control over if structures should generate and/or where.

This PR is required for simple, complete, and effective structure generation.

maxanier committed:
Add support for retroactively generating structures

maxanier committed:
Address issues

maxanier commented:
Added interface to opt-in to retroactively generating structures. Namechange suggestions welcome.
Also added discussed comment and fixed/changed spawning

maxanier committed:
Disable test

mezz commented on src/main/java/net/minecraftforge/common/MapGenStructureManager.java:
```diff
@@ -0,0 +1,184 @@
+    @Nullable
+    public BlockPos getNearestStructurePos(World worldIn, String structureName, BlockPos position, boolean findUnexplored)
+    {
+        if (structureMap.containsKey(structureName))
```
You can avoid this `containsKey` call and just do `get` with a null check

mezz commented on src/main/java/net/minecraftforge/common/MapGenStructureManager.java:
```diff
@@ -0,0 +1,184 @@
+     */
+    public boolean isInsideStructure(World worldIn, String structureName, BlockPos pos)
+    {
+        if (structureMap.containsKey(structureName))
```
avoid this `containsKey`

mezz commented on src/main/java/net/minecraftforge/common/MapGenStructureManager.java:
```diff
@@ -0,0 +1,184 @@
+
+    public boolean retroGenerateStructures(World world, Random rand, ChunkPos chunkPos, Chunk chunk){
+        boolean flag=false;
+        for(MapGenStructure structure : structureMap.values()){
```
Please fix formatting in this method, it's all strange.

maxanier replied:
Sorry about that, I keep forgetting that I disabled the auto-formatter for Forge

mezz commented on src/main/java/net/minecraftforge/common/MapGenStructureManager.java:
```diff
@@ -0,0 +1,184 @@
+     */
+    protected static void addStructureName(String structure)
+    {
+            allStructureNames.add(structure);
```
strange indentation here

mezz reviewed (changes requested):
Looking pretty good. 
Is it possible to disable/replace vanilla structures with this system? If not, can wrappers be made for vanilla things?

pau101 commented:
@mezz If a modder wanted to disable or otherwise control vanilla structures the existing `InitMapGenEvent` event could be used.

maxanier committed:
Updated style and avoid contains check

pau101 commented on patches/minecraft/net/minecraft/world/gen/ChunkGeneratorOverworld.java.patch:
```diff
@@ -100,3 +113,51 @@
+             flag |= this.field_185980_B.func_175794_a(this.field_185995_n, this.field_185990_i, new ChunkPos(p_185933_2_, p_185933_3_));
+         }
+ 
++        flag |= this.structureManager.retroGenerateStructures(this.field_185995_n, this.field_185990_i, new ChunkPos(p_185933_2_, p_185933_3_), p_185933_1_);
```
Should only occur if `mapFeaturesEnabled` is true

maxanier replied:
Why? The other methods generate are called regardless of this value, so a mod might decide to generate something anyway (even though not recommended).

jabelar commented:
> @jabelar for full structure support there are a handful of methods which a custom structure needs ties into, which is what this pr is doing. For example the locate command, cartographer maps, natural mob spawning, and usage in the location predicate for advancement triggers. Doing a wrap of an existing generator like you mention only enables generation of your own structure but it can not support the aforementioned additional features.

Actually, the wrapper works fine for all that because it is called exactly at the same time in the same way, honoring things like is map features enabled. However, I admit it is slightly hacky -- basically I take over one of the map gens in the existing chunk generator and call all my map gens from there. The locate command I just register my own, and so on. But again admittedly hacky.

> In the case for ChunkGeneratorOverworld in regards to mapFeaturesEnabled which require AT or reflection,

That is not true, the value is set and accessible directly from world.getWorldInfo().

But I'm not here to quibble -- it is cool if you're doing all that in a more managed way.

LexManos commented:
> If a modder wanted to disable or otherwise control vanilla structures the existing InitMapGenEvent event could be used.

Yes, but if we move things to a more standard ordered system, then we can cut down the special cased vanilla stuff and potentially remove the InitMapGenEvent in favor of a system that treats both vanilla and moded the same.

pau101 commented:
I'm all for streamlining the structure handling and removing InitMapGenEvent. It was mentioned by name for sake of discussion.

Alpvax commented:
How does this interact with the superflat rules? Is it possible to enable/disable generation of mod structures using them?

maxanier commented:
> Yes, but if we move things to a more standard ordered system, then we can cut down the special cased vanilla stuff and potentially remove the InitMapGenEvent in favor of a system that treats both vanilla and modded the same.

If we would want to (almost completely) streamline the vanilla structures with the modded system, this would mean a much bigger patch size and a bigger PR. 
Not sure what you have in mind, but we could move the interfaces methods to MapGenStructure, move  the vanilla specific code, to the appropriate Vanilla MapGenStructures (spawning, retrogen, ...) and add a new way to register and substitute MapGenStructures.

However, I am not sure if that is reasonable. If so, I would prefer doing two separate PRs (this one including preparation for Vanilla streamlining and a breaking/1.13 PR which moves the Vanilla things over).

Maybe I got your suggestion wrong though. Let me know

maxanier commented:
> How does this interact with the superflat rules? Is it possible to enable/disable generation of mod structures using them?

As far as I can remember I have excluded the ChunkGeneratorFlat because it has limited use for modded in my opinion, but yes I will have another look and add it too. Should be straightforward.

LexManos commented:
It wouldn't need a too much larger patch. Just would require wrapping the vanilla generator objects in wrappers and registering them.

Something like:
```
builder.register(new ResourceLocation("minecraft:villages", new VanillaWrapper(this.villageGenerator, () -> this.settings.useVillages));
VanillaWrapper{
  generate(){
    if (enabledFunc.call())
      vanilla.generate()

```

maxanier commented:
Good point, using wrappers, we could keep the patch size rather small (just have to remove all the direct Vanilla calls). 

Guess there are different ways of implementing this, let me explain my initial idea to see if I am on the same page as you:

- Have a new interface (e.g. IForgeMapGenStructure or something more creative), which defines all the methods that have to be called by StructureManager (for generation).
- Any structure to register must implement this (which means they don't have to extend MapGenStructure anymore even though there is little reason not to)
- Vanilla structures stay unmodified in the respective chunk generator (so they can be accessed by mods as before), but are also added to the StructureManager using a simple wrapper (might possible with just one wrapper class, but might need some more specific subclasses e.g. for the ocean monument, will have to see).
- Remove direct calls to vanilla structures as they are called indirectly via the wrappers added to the StructureManager.
- Add some way to substitute registered structures in the StructureManager to be able to replace/modify vanilla behaviour.

Still not sure if it is worth the effort

One thing I am not sure about is added interfaces and obfuscation:
What happens if I create an interface including a method with the exact same signature as a method in MapGenStructure and then implements this interface in a subclass of MapGenStructure. Without obfuscation any calls to the interface method would go to MapGenStructure, but not sure what happens with obfuscation

LexManos commented:
Obfuscation is an issue. You'd want to either have different method names/signatures. Or instead of having MapGenStructure implement the interface, have the wrapper do it.
Removing the vanilla function calls is simple. Only large chunk would the Overworld gen but that can be a simple if (false) patch set.

As for the effort vs worth. That is up to you guys. If its not worth the effort to do right then its not worth going into Forge.

maxanier commented:
Mhh, that means we would need to have a wrapper even for the modded structures :/ ok, but not the nicest solution. Have not come up with a better solution yet, though. Anyone else got a suggestion?

About doing things right: 
I would like to state that the current PR does it "right".
It is a clean and simple system allowing a easy registration of new structures while keeping the modifications to Vanilla small.  
The Vanilla structures have to be modified via a different system, yes, but the goal is a different there. We do want to substitute Vanilla structures with modified ones, but we don't want to add "new Vanilla" structures nor do we want to substitute structures added by mods. Therefore it is reasonable to have two different systems.

Merging both systems (with slightly different goals) is good too, of course. But it does not get rid of Vanilla "special casing" it is just moved somewhere else, making it harder to maintain.
So I would say these are both right approaches, just with different advantages and disadvantages.

jabelar commented:
Hey in the release notes for a 1.13 snapshot they mention they are rewriting world generation "from the ground up". Do we know enough to know whether investing in forge hooks in world gen for 1.12 is worth it until we see what they've done?  See: https://www.minecraftforum.net/news/60668-minecraft-1-13-snapshot-18w06a

Alpvax commented:
> As far as I can remember I have excluded the ChunkGeneratorFlat because it has limited use for modded in my opinion, but yes I will have another look and add it too. Should be straightforward.

Worth supporting it, just because it is hardly used doesn't mean that no-one wants to. If forge structures support it, custom worldtypes might start to use a similar system for enabling/disabling generation features.

> Hey in the release notes for a 1.13 snapshot they mention they are rewriting world generation "from the ground up".

I agree that we at least need to know as an overview what has changed, if not specifics.

maxanier commented:
To be honest I lost motivation for this PR by now.
Maybe 1.13 will allow of something similar, let's see. Otherwise there are always some other dirty tricks to make things like this work.

Niko-sk2x commented:
I was looking at 1.13 and it will almost definitely allow to do this.

mezz commented:
Thanks for working through this anyway @maxanier.
It's always nice when fairly complicated changes like this are done by vanilla directly.
